UI: Reflect search in URL

jh3y opened the pull request:
Issue: #16287

Integrate way to reflect search("filter") in URL bar in order to share the URL for hosted Storybook instances.

## What I did
- Grab the `Sidebar` `Search` `initialQuery` from `URLSearchParams` and use the `filter` value.
- Tap into `Downshift` render to update the URL bar with `window.history.replaceState`. 

## How to test

- [x] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

Tests are under `Search.test.js`. Tests populating the input field in addition to update `window.location` via mocked `window.history` 👍 
<!--

Everybody: Please submit all PRs to the `next` branch unless they are specific to the current release. Storybook maintainers cherry-pick bug and documentation fixes into the `master` branch as part of the release process, so you shouldn't need to worry about this.

Maintainers: Please tag your pull request with at least one of the following:
`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

-->

jh3y commented:
Can we add a `hacktoberfest-accepted` label to this PR? 🙏

domyen commented:
Thanks for the PR @jh3y! I tagged in @ghengeveld who built search and knows about the URL structure.

jh3y commented:
Rad! Thanks @domyen 🙏 
I don't think I've broken anything but we'll see 🤞 😅

yannbf commented:
Hey @ghengeveld could you please take a look at this whenever you've got time?

jh3y commented:
Bumping this so `stale` doesn't close it for inactivity ✨ 

Happy New Year! 🎉

yannbf commented:
> Bumping this so `stale` doesn't close it for inactivity ✨
> 
> Happy New Year! 🎉

Hey @jh3y thanks a lot for your patience! I'll make sure this PR is reviewed this week

ghengeveld commented:
URL params should not be handled by UI component. Instead the `filter` (or perhaps `query` or `search`) param should be handled similarly to other UI query params like `shortcuts`. See `initialUrlSupport` in [url.ts](https://github.com/storybookjs/storybook/blob/next/lib/api/src/modules/url.ts). A `initialQuery` property on the [UI](https://github.com/storybookjs/storybook/blob/2f83d1b7a1cb419f2add231f4c631d5b4b8659be/lib/api/src/modules/layout.ts#L28) state object makes sense here.

jh3y commented:
I'll take another look at this. Thanks for the links @ghengeveld 🙏  Will take me a minute to refresh my context knowledge for this again 😄

jh3y commented:
Assuming the `filter` string is grabbed in `url.ts`. How do you grab that piece of state within a UI component @ghengeveld? Or do we need to pass it down through one of the components further up in the stack?

ghengeveld commented:
Great, thanks. It's appreciated!

jh3y commented:
Updated @ghengeveld 🚀 

- Passing through a `filter` key in the `ui` key of `state`.
- Updated the tests with a mocked return for `useStorybookState`.

Only thing is that the core tests are timing out. Not sure if it's a case of giving them a rerun as it's related to something with Vue 🤔 

Thanks! ʕ´•ᴥ•`ʔ

ghengeveld commented on lib/ui/src/components/sidebar/Sidebar.tsx:
```diff
@@ -109,6 +111,8 @@ export const Sidebar: FunctionComponent<SidebarProps> = React.memo(
     const isLoading = !dataset.hash[DEFAULT_REF_ID].ready;
     const lastViewedProps = useLastViewed(selected);
 
+    console.info(state, 'STATE');
+
```
```suggestion
```

jh3y replied:
Thanks for catching that! 🙌

ghengeveld reviewed (approved):
Nice work! Maybe setting the URL should not happen in the sidebar, but at least we're not reading from it there. @shilman wdyt?

shilman reviewed (changes requested):
This change looks great @jh3y! However, it highlights a bug in the existing search UI, that I think we should also fix in this PR.

### Current behavior
1. I search, it shows filtered search results
2. I click on a result and it navigates to that story returns to the normal sidebar UI
3. However, it leaves my query in the search UI, which I think is a minor bug
4. I share the URL and it links the story I'm currently looking at

### New behavior
1. same
2. same
3. It leaves my query in the search UI and also the URL query parameters
4. I share the URL and it links to the search results, even though I'm viewing a story by this point in the flow

@ghengeveld do you agree this is a bug?

@jh3y any chance you can fix this behavior as part of this PR?

jh3y commented:
Yeah, I see this @shilman 👍 

Looking at it now. So would need to capture the click event on side bar items and the focus event of the search UI I imagine. Got it open, I'll dig in 👍 

Thanks for pointing it out! 🙏

shilman commented:
Thanks so much @jh3y !! 

@ghengeveld would know the best way to fix this. In addition to UI level stuff there's also an event emitted on Storybook's channel (that communicates between the preview iframe and the manager UI) that might also be appropriate to listen to. I'm not so familiar with the sidebar code, alas.

jh3y commented:
No problem @shilman 🙏 

I think on first glance, we can hook into the navigation helpers and append the `filter` on where applicable. That would catch it a high level I believe. Will keep poking. Don't think it will take much to add 👍 🤞

ghengeveld commented:
The behavior of keeping the query in the search input after selecting a result is intentional, so that when you press `/` again you will be able to either replace the query (it gets selected automatically), or extend it by pressing `→` and typing more. I would expect the URL to contain the query (filter) regardless of whether the search pane is open right now (so selecting a result does not remove the filter param from the URL). Obviously when the query input is empties, the `filter` param should be removed altogether.

I think the intended behavior for the `filter` URL param is to open the search panel with that query when you open a Storybook at that URL, right? In that case it makes sense to set the search panel to be open when a `filter` param is provided (and not empty). I think you can just do this in Search.tsx:

```jsx
<Downshift<DownshiftItem>
  initialInputValue={initialQuery}
  initialIsOpen={!!initialQuery}
  ...
>
```

jh3y commented:
Yeah, I'm just trying to find the best spot to keep the URL in sync with this 🤔 

Would it be because the `store` needs updating when the filter is changed? Then I can use `store.getUrlState` to correctly keep the `url` in sync?

jh3y commented:
As in, what's the best way to emit an event from the search input and catch that to update it in `state.ui`? 🤔

jh3y commented:
Think there was a better way to approach this potentially and I've made some updates that hook into using the `api` to update `customQueryParams` and use them as it works across the board like that as expected 👍

shilman commented:
@ghengeveld So, as a user, I should know to clear the search manually before sharing the URL to my colleague? That feels like a bug to me.

See my comment above: https://github.com/storybookjs/storybook/pull/16527#pullrequestreview-872954977

jh3y commented:
I'm totally open to whichever way we want to do this 😄 

I can revert the latest changes if necessary or update these to filter out certain `customQuery` params 🤔

shilman commented:
@jh3y thanks for being flexible! i'm going to stick to my guns on this, barring an alternative proposal that addresses my concern. 😅

jh3y commented:
No problem @shilman 🙏 

Can we outline the "exact" behavior we expect then? Or is it a case of it did kinda work how we wanted it to? I can see the pros and cons of both approaches. The implementation is "clearer" as-is I guess.

I'm interpreting it that we only want the `customQuery` to be populated in "some" scenarios for the URL. But what are those flows?

jh3y commented:
Giving this one a bump as it's been 7+ days ✊ 

I would love to make a contribution into the project 🚀

ghengeveld commented:
Okay, let's recap. So far and as per my earlier suggestion, you've implemented the following behavior:
- Typing in search field adds `filter` parameter to URL.
- Closing the search (e.g. by selecting a result) keeps the last query in the field, as well as in the URL
- Opening a Storybook with a `filter` parameter in the URL will fill and open the search with that query.

The above is problematic because, like @shilman described, users tend to share their Storybook URLs with co-workers, but that URL might still contain a `filter` param, even if it isn't their intention to link to a particular search query (with open search panel). People don't tend to manually clear the search field, so this is a very likely scenario. It's also not obvious that sharing such a URL will end up opening the search panel when the URL is used.

I see a couple of options to rectify this problem:
- Remove the `filter` param when closing the search panel (but keep the query in the search field). Set it again when the search panel is reopened.
- Keep the `filter` param, but add another param to track the "open" state of the search panel. This way one can share a URL with an initial query, as well as with the panel open or closed.

Contrary to my earlier suggestion, I think the first option makes most sense now. I don't see much value in being able to link to a Storybook with a prefilled search field, but with the search panel closed. It's much more likely to be a mistake than to be the intention of whoever shared the URL. If you want to share a particular query, it makes sense to show the results of that query when you follow the URL.

jh3y commented:
@ghengeveld @shilman – Found a moment to revisit this 👍 

What approach do we think would be best here? I was thinking of having some kind of `global` state that tells us whether the search results are currently being shown? And then hooking into that within the `url` module perhaps?

I'm not sure how else to trigger it or whether I'm overthinking it in the sense of we can set the query params of `filter` to `null` when not required? However, when I've tried that, it doesn't update as I'd expect it to.

```
syncUrlToFilter(isBrowsing ? null : filter);
api.setQueryParams({ filter: isBrowsing ? null : filter });
```

The call to `setQueryParams` with a value of `{ filter: null }` doesn't delete it from `customQueryParams` as I'd expect 🤔

yannbf commented:
@shilman @ghengeveld can you check this out? Thanks!

ndelangen commented:
@ghengeveld please take a look at this?

@jh3y Thank you for keeping with this PR, I have to defer to @ghengeveld and @MichaelArestad for these types of UX changes. I like the feature, but I'm a bit indifferent about the implementation. I'm sure @ghengeveld has ideas on this.

ndelangen commented:
@jh3y Looks like some CI things are failing after i merged in the base branch. Do you recon you could have a look to see if you could fix that? Thank you!

ghengeveld commented on lib/ui/src/components/sidebar/Search.tsx:
```diff
@@ -324,10 +324,12 @@ export const Search = React.memo<{
           let results: DownshiftItem[] = filter ? getResults(filter) : [];
 
-          syncUrlToFilter(filter);
+          syncUrlToFilter(isBrowsing ? null : filter);
+          api.setQueryParams({ filter: isBrowsing ? null : filter });
```
I think we don't even need `syncUrlToFilter` and can just rely on `api.setQueryParams` to do the replaceState for us. The only thing is the value cannot be `null` because `setQueryParams` will ignore those. Probably you can just set it to `undefined`.

ghengeveld commented on lib/ui/src/components/sidebar/Search.tsx:
```diff
@@ -303,10 +324,14 @@ export const Search = React.memo<{
+          const filter = inputValue ? inputValue.trim() : '';
+          let results: DownshiftItem[] = filter ? getResults(filter) : [];
+
+          syncUrlToFilter(isBrowsing ? null : filter);
+          api.setQueryParams({ filter: isBrowsing ? null : filter });
```
I think we don't even need `syncUrlToFilter` and can just rely on `api.setQueryParams` to do the `replaceState` for us. The only thing is the value cannot be `null` because `setQueryParams` will ignore those. Probably you can just set it to `undefined`, but I'm not exactly sure about its behavior.

jh3y replied:
Are saying removing `syncUrlToFilter` will satisfy the feature then? I'm sure I put it in for a reason but it seems so long ago now 😅 

I can take it out and try to get the CI playing ball again 👍

jh3y commented:
> @jh3y Looks like some CI things are failing after i merged in the base branch. Do you recon you could have a look to see if you could fix that? Thank you!

Yep! I'll try and get this sorted when I get a minute. A lot of life stuff has changed since opening this PR 😅

ndelangen commented:
good life stuff, I hope!

ndelangen commented:
@jh3y @ghengeveld Is this PR salvageable? Can this be merged still?

It's likely that the code has changed a lot since?

jh3y commented:
Ooof – Yeah, if there was ever a time to bring this one back, it's now! Is it merely solving the conflict here and it's "Good to go!"?

ndelangen commented:
I don't know... I'm hoping @ghengeveld does..

ghengeveld commented:
Yes, I think we were in a state where this was mergeable, so if conflicts are resolved and everything behaves as discussed, it has my blessing 🙌

jh3y commented:
Took a little tinkering... but... this _should_ work now after resetting the branch against the latest `next` 🤞 The conflicts are gone!

I may need a little guidance if the tests are no longer passing. But, they did pass previously, promise 😅 

@jh3y

jh3y commented:
Ahh – As expected. I was previously mocking the `api` but that doesn't seem to be working anymore. Any ideas as to what might have changed there?

ndelangen commented:
@jh3y I think I resolved the merge-conflicts, but I think there's a bit of polishing up to do.
A bunch of things got moved in the storybook monorepo during this PR being open (absolutely not your fault btw!)

One of which was `@storybook/api` being renamed to `@storybook/manager-api` for clarity.

Another thing that happened was the UI code was moved to a different directory.

And third but not least, typescript strictness was increased.

I hope with this background you might be able to polish the last remaining issues of this PR, and then I can merge it.

Sorry it has taken ages, If you need any help getting this green and merged, I'd be happy to do that in a pairing session, schedule a meeting here: 
https://calendly.com/chromaui

ndelangen commented:
Is anyone interested in taking on this PR, bringing it up to dat, fixing type-issies and make unit-tests pass?

jh3y commented:
I should be able to get it sorted still. Just haven't had a moment to look at it. And I get the feeling it's not going to be a sit down for five minutes and get it done kind of job 😅 

Thanks for the information up above! 🙏

ndelangen commented:
Ok, I understand, thanks for letting me know ❤️

jh3y commented:
Having a look at this now 👍 

It's probably going to be less effort to create a new branch off `next` and force push onto this one with the changes in as it's only a handful of files 👍 

Jus' bootstrapping a new clone. Forgot this takes a minute 😅

jh3y committed:
feat: UI: reflect search in url #16287

jh3y commented:
Finally! 🎉 😅 

It may require some extra type work but it passed the linter so it _should_ be good 🤞 

Happy to make any changes though. I ran everything local and it looked like it was doing the right things 👍 

- Jhey

jh3y commented:
This close 🤏 

I can't get `prefills input with search params` to pass for some reason.

Running the test trying to render `SidebarStories.Simple` fires an error about not being able to read property `docsMode` of `undefined`. I'm assuming that I'm missing a mock of something in order for it to render in the test environment 🤔

jh3y commented:
It's close 🤏 

I can't quite get the last test to pass `prefills input with search params`

It complains about reading property `docsMode` of `undefined`. I'm assuming I'm maybe missing a mock of something in order for `SidebarStories.Simple` to render in the test env.

That test _might_ not be required though if we say it's handled by testing the render of `Search` based on its `props`. However, this test is meant to test if the URL gets picked up in `customQueryParams` and passed through to `Search`. It does when running it.

jh3y commented:
Givin' this a little Monday bump as I don't want to fall too behind against `main` 😁

jh3y commented:
🔔

jh3y commented:
Shall we remove that failing test @ndelangen? 🤔 I'm not sure how to mock out that scenario and it might not even be required.

– Jhey

jh3y commented:
Had another swing at getting the tests and CI to pass on this with no luck unfortunately 🤔 
The feature is functional and unit tests pass locally.

But, I'm not 100% how to make it happy when those CI tests pass. It doesn't like the mocking outside of unit tests.

ndelangen commented:
@jh3y I'll investigate and try to merge after the 7.0 release.

ndelangen committed:
Merge branch 'next' into pr/jh3y/16527

ndelangen committed:
Add initialQuery prop to Sidebar component

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

43081j commented:
i caught this up from `next` FYI, then realised a few things were broken (conflicts, a bug, etc).

not realising i was still mid-merge, i fixed those and reworked the tests (so it shows up as a merge commit but we'll squash it anyway i imagine when it lands).

TLDR of what I changed:

- the tests now test `<Search>` directly (unrelated to the sidebar etc)
- dropped the `testing-library` dependency so we resolve to the same one the rest of the monorepo uses (this was causing a test failure otherwise, for whatever reason, old version i guess)
- we no longer update the URL or call `setQueryParams` on render, only now calling it on user input
- moved the main logic out of `onInput` and into downshift's own change handler (resolves a bug where the `x` button would clear the input but not change the url)

jh3y commented:
> i caught this up from `next` FYI, then realised a few things were broken (conflicts, a bug, etc).
> 
> not realising i was still mid-merge, i fixed those and reworked the tests (so it shows up as a merge commit but we'll squash it anyway i imagine when it lands).
> 
> TLDR of what I changed:
> 
> * the tests now test `<Search>` directly (unrelated to the sidebar etc)
> * dropped the `testing-library` dependency so we resolve to the same one the rest of the monorepo uses (this was causing a test failure otherwise, for whatever reason, old version i guess)
> * we no longer update the URL or call `setQueryParams` on render, only now calling it on user input
> * moved the main logic out of `onInput` and into downshift's own change handler (resolves a bug where the `x` button would clear the input but not change the url)
> * use vitest instead of jest

Legend 🙏 A lot has changed since first opening this PR. Looking back on this, I got it rebased at whatever stage and then couldn't seem to get the various tests passing at that time.

Thanks for taking a look!

JReinhold commented on code/ui/manager/src/components/sidebar/__tests__/Search.test.tsx:
We should be able to have these tests as stories+play functions instead of Vitest's, allowing us visually inspect the tests more easily.

The mock `@storybook/manager-api` can be achieved with a custom `ManagerContext.Provider`-decorator which we already do a couple of other places:

https://github.com/storybookjs/storybook/blob/next/code/ui/manager/src/components/sidebar/Sidebar.stories.tsx/#L47-L70

43081j replied:
i've moved them, although i'm not sure how we can achieve an equivalent test of:

- set the query params
- _then_ load the story
- observe that it already had the search filled in

also the `check` task is unhappy but ill sort that some other time

JReinhold commented on code/ui/manager/src/components/sidebar/Search.tsx:
```diff
@@ -28,6 +28,8 @@ const { document } = global;
 
 const DEFAULT_MAX_SEARCH_RESULTS = 50;
 
+export const FILTER_KEY = 'filter';
```
I have a slight long-term concern here that this could conflict with the `experimental_setFilter` API, which we someday might want to expose in the URL too?

Could this be `'search'` instead maybe?

43081j replied:
renamed to `search` 👍

JReinhold reviewed (commented):
Impressive work bringing this up to date!!

43081j committed:
test: rework tests into stories

JReinhold commented on code/ui/manager/src/components/sidebar/Search.stories.tsx:
```diff
@@ -20,76 +20,105 @@ const getLastViewed = () =>
+              off: () => {},
+              getShortcutKeys: () => ({ search: ['control', 'shift', 's'] }),
+              selectStory: () => {},
+              setQueryParams: (params: Record<string, string>): void => {
+                queryParams = params;
+              },
+              getQueryParams: () => queryParams
```
here you could use a mock functions instead of the top-level `queryParams`, and then mock the return value per story and assert on the `setQueryParams` as you want.

```
import { fn } from '@storybook/test';

...

const setQueryParamsMock = fn().mockName('setQueryParams');
const getQueryParamsMock = fn().mockName('getQueryParams');
```

```suggestion
              setQueryParams: setQueryParamsMock,
              getQueryParams: getQueryParamsMock
```

then in tests:

```ts
const MyStory = {
  ...,
  beforeEach: () => {
    getQueryParamsMock.mockReturnValue({ whatever });
  },
  play: () => {
    expect(setQueryParamsMock).toHaveBeenCalledWith({ ... });
  }
}
```

The `beforeEach` hook is being introduced in 8.1 as we speak so it makes sense that you wouldn't know about that yet. https://storybook.js.org/docs/8.1/writing-stories/mocking-modules#using-mocked-modules-in-stories

43081j replied:
aha i didn't realise we had a `beforeEach`! all good, i'll update 👍

43081j committed:
chore: use test functions in stories

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

yannbf commented:
Amazing work!! The story failures in Chromatic are not related to this PR. I just made another PR that fixes them so once that's merged, let's just get the commits in here to fix that.
https://github.com/storybookjs/storybook/pull/27145

JReinhold committed:
Merge branch 'next' into reflect-search-in-url-16287

43081j commented:
the current build failure seems to be mostly due to having stories with args.

im slightly baffled by this because the `Sidebar` stories do the same and work fine 🤔 

both `Sidebar.stories.tsx` and `Search.stories.tsx` have exactly this line:

```ts
export const Simple: Story = {};
```

but the search stories fail type checking..

43081j committed:
fix: add correct types for story props

43081j committed:
chore: run format

yannbf committed:
Merge branch 'next' into reflect-search-in-url-16287

yannbf commented:
@43081j that other PR was merged so I updated this branch (and fixed a small merge conflict), it should all be good now!

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

43081j commented:
have caught this up from `next`. any chance we can find time to merge it some day soon? 👀

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

ndelangen committed:
Merge branch 'next' into pr/jh3y/16527

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

JReinhold commented on code/core/src/manager/components/sidebar/Search.tsx:
```diff
@@ -272,9 +278,26 @@ export const Search = React.memo<{
+      api.setQueryParams({
+        [FILTER_KEY]: isBrowsing ? undefined : inputValue,
+      });
+      const params = new URLSearchParams(window.location.search);
+      if (window.history.replaceState) {
+        if (inputValue) {
+          params.set(FILTER_KEY, inputValue);
+        } else {
+          params.delete(FILTER_KEY);
+        }
+        const paramsString = params.size > 0 ? `?${params.toString()}` : '';
+        window.history.replaceState({}, '', paramsString);
+      }
```
this isn't needed (maybe it was once), I'm pretty certain `api.setQueryParams()` already does this.

```suggestion
```

43081j replied:
im not sure it actually does unless im just looking in the wrong place. i expected it to for sure but this doesn't seem that way:

https://github.com/storybookjs/storybook/blob/6388f93b1f3517b8b0dad3dbd05aa5d5313ee5ae/code/core/src/manager-api/modules/url.ts#L192-L208

JReinhold replied:
You're looking at the right place, but the code flow is a long and winding road 😅

1. `setQueryParams` emits the `UPDATE_QUERY_PARAMS` event on line 206
https://github.com/storybookjs/storybook/blob/6388f93b1f3517b8b0dad3dbd05aa5d5313ee5ae/code/core/src/manager-api/modules/url.ts#L192-L208

2. The `PreviewWithSelection` instance listens for that event 
https://github.com/storybookjs/storybook/blob/6388f93b1f3517b8b0dad3dbd05aa5d5313ee5ae/code/core/src/preview-api/modules/preview-web/PreviewWithSelection.tsx#L107

3. ... and calls `setQueryParams` on the `SelectionStore` instance. https://github.com/storybookjs/storybook/blob/4419ab2fce100d287ec8b5aa01db9c81151f0e0d/code/core/src/preview-api/modules/preview-web/PreviewWithSelection.tsx#L235-L237

4. The `selectionStore` happens to be a `UrlStore` that does `history.replace` in the end. https://github.com/storybookjs/storybook/blob/4419ab2fce100d287ec8b5aa01db9c81151f0e0d/code/core/src/preview-api/modules/preview-web/UrlStore.ts#L106-L110

43081j replied:
Aha thanks a lot for the explanation, makes sense to me!

JReinhold replied:
You were right I was wrong!
The `PreviewWithSelection` that listens for the event is on the preview-side (the iframe that shows the story), so `setQueryParams` would only set the query params in the iframe, not in the actual top-level manager.

I changed it to use `applyQueryParams`, which fixes it.

JReinhold commented on code/yarn.lock:
There shouldn't be a diff in `yarn.lock` based on what I see, probably just a merge commit gone wrong?

JReinhold commented on code/core/src/manager/components/sidebar/Search.stories.tsx:
```diff
@@ -23,77 +22,101 @@ const getLastViewed = () =>
+    getLastViewed: (): Selection[] => [],
+    children: () => <SearchResults {...noResults} />,
+  },
+  render: ({ children, ...args }) => {
+    return <Search {...args}>{children}</Search>;
+  },
```
This looks like the default render-function to me, ie. `<Search {...args} />` implicitly has `children` too.

```suggestion
```

JReinhold commented on code/core/src/manager/components/sidebar/Search.stories.tsx:
```diff
@@ -23,77 +22,101 @@ const getLastViewed = () =>
   component: Search,
   title: 'Sidebar/Search',
   parameters: { layout: 'fullscreen' },
+  args: {
+    dataset,
+    getLastViewed: (): Selection[] => [],
+    children: () => <SearchResults {...noResults} />,
+  },
```
Based on this line https://github.com/storybookjs/storybook/pull/16527/files#diff-29c94008593fc41c988e602ce18ec169d7212632502bff1b0334c2f92dac12d8L44 we probably also want a mock function for `clearLastViewed`

```suggestion
  args: {
    dataset,
    getLastViewed: (): Selection[] => [],
    clearLastViewed: fn().mockName('clearLastViewed'),
    children: () => <SearchResults {...noResults} />,
  },
```

43081j replied:
the component doesn't have a `clearLastViewed` so this causes a build error. maybe it was in there incorrectly to begin with?

JReinhold replied:
aha I see, sounds about right.

JReinhold commented on code/core/src/manager/components/sidebar/Search.stories.tsx:
```diff
@@ -23,77 +22,101 @@ const getLastViewed = () =>
+    const clearIcon = await canvas.findByTitle('Clear search');
+    await userEvent.click(clearIcon);
+
+    expect(setQueryParams).toHaveBeenCalledWith({ search: null });
```
Based on https://github.com/storybookjs/storybook/pull/16527/files#diff-39f5915dc511a2d421234f378dc2a5efbaef09acfe73d81f95e440838c1326f1R286 this should be `undefined` instead of `null`, which is also evident by the failing test: https://www.chromatic.com/test?appId=635781f3500dd2c49e189caf&id=66c22bd5c11b883b85898ff9

JReinhold commented on code/core/src/manager/components/sidebar/Search.tsx:
```diff
@@ -2,13 +2,15 @@ import React, { useCallback, useRef, useState } from 'react';
 
 import { IconButton, TooltipNote, WithTooltip } from '@storybook/core/components';
 import { styled } from '@storybook/core/theming';
-import { global } from '@storybook/global';
```
This requires that line 32 below is then changed to

```
const { document } = globalThis;
```

(which is great, btw)

See the published Storybook failing to load https://635781f3500dd2c49e189caf-ytrwlnbeyv.chromatic.com/

JReinhold replied:
Sorry, looks like there are still some `global` references left, I missed those.

https://github.com/storybookjs/storybook/blob/6ad7a0b4bc45830065fa1aa1139527d0e8bc46fe/code/core/src/manager/components/sidebar/Search.tsx#L174-L175

JReinhold commented on code/core/src/manager/components/sidebar/Search.stories.tsx:
```diff
@@ -23,77 +22,101 @@ const getLastViewed = () =>
+};
+
+export const LastViewed: Story = {
+  args: {
+    getLastViewed,
+  },
```
These lines have been removed, is that on purpose?

```suggestion
  args: {
    getLastViewed,
    children: ({ query, results, closeMenu, getMenuProps, getItemProps, highlightedIndex }) => (
      <SearchResults
        query={query}
        results={results}
        closeMenu={closeMenu}
        getMenuProps={getMenuProps}
        getItemProps={getItemProps}
        highlightedIndex={highlightedIndex}
      />
    )
  },
```

maybe this is just the equivalent of the default `    children: () => <SearchResults {...noResults} />,` arg?

43081j replied:
from what i remember, its because the default handles it. none of the others specify `children` other than the default

JReinhold reviewed (changes requested):
Still some work left to do, but feels like some rather simple changes needed :) 

I can help get this over the finish line, if needed.

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

43081j committed:
chore: clean up some PR feedback

JReinhold committed:
Fix global references

JReinhold committed:
Merge branch 'next' into reflect-search-in-url-16287

JReinhold committed:
use applyQueryParams instead of setQueryParams

JReinhold commented:
I tested it out and it looked to be working, apart from a minor quirk. See the video for what I mean:

https://github.com/user-attachments/assets/aafff355-7773-4ea9-bc26-c42c5ff3868a

To make it clear what's going on with the query param when a story is selected:

1. starts as the input value, "add"
2. changes to the story name, "parameters"
3. navigates to the new story and clears the query param
4. re-adds the story name "parameters" as the search query param again
5. finally clears the search query param

I'm pretty sure this is all because of how Downshift triggers the events, as this is also visible from the actions being triggered by the Clearing story in the Search component:

![image](https://github.com/user-attachments/assets/72d840aa-aca3-4272-8d0d-817460142ed4)

43081j commented:
that's a slight pain 😬  do you have any suggestions of what we could do about it? if anything

I'm out travelling a lot these next few months so any help to finish this off would be much appreciated

ghengeveld commented:
@43081j Would debouncing help here perhaps?

SpanishPear commented:
👋 I played around with this for a little bit, mostly out of curiosity, to see if I could help resolve this! The good news, is that I can pretty much always replicate that **selecting** an item, will set the params to that item, before quickly being unset.

E.G - in the recording above, that meant:

1.  `addons-actions-config--action-a&search=add` 
2. **click** on addon-a11y-parameters
3. `addons-actions-config--action-a&search=parameters` (❗)
4. `addon-a11y-parameters--options` (from 2)
5. `addon-a11y-parameters--options&search=parameters` (❗)
6. `addon-a11y-parameters--options`

This leaves us with some questions.
- Q) Why is selecting a story changing the selected search value? Shouldn't only typing (a.k.a [this ](https://github.com/storybookjs/storybook/pull/16527/files#diff-39f5915dc511a2d421234f378dc2a5efbaef09acfe73d81f95e440838c1326f1R281)change to `onInputValueChange`) trigger the queryParam update? Clicks should be captured by the Downshift Reducer ([ref](https://github.com/storybookjs/storybook/blob/next/code/core/src/manager/components/sidebar/Search.tsx#L291)) right?
  - A) Unfortunately, it seem, this is a known and unsolved bug inside Downshift, https://github.com/downshift-js/downshift/issues/759
- Q) Why does the URL flicker from `&search=parameters`(3) to no parameters (4) to `&search=parameters` (5)
   -  A) I believe this is intended, as storybook saves and restores parameters upon navigating stories via it's API store. You can see some of the implementation for get/set query params [here](https://github.com/storybookjs/storybook/blob/next/code/core/src/manager-api/modules/url.ts#L186), specifically how it pulls/pushes from a central store :) 


Unfortunately, I think that means this PR will either:
 - Accept the "jank" and merge after resolving conflicts
 - Figure out an alternative workaround to the above downshift bug
 - Rewrite away from using Downshift 😢 
 - Abandoned/Declined

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

43081j committed:
test: add title back to regain passing test

43081j committed:
Merge branch 'next' into reflect-search-in-url-16287

yannbf committed:
Merge branch 'next' into reflect-search-in-url-16287

yannbf commented:
Hey everyone, I think it's time to close the loop here. Although the feature is really good and would be a nice addition to Storybook, we haven't focused on this PR for the longest time. There are a few implications as can be seen in the discussions and therefore this PR needs to be revisited, we need to account for a11y, UX, etc. I think the best course forward would be to close this PR, and write this as a feature request that we can potentially include in the plans of Storybook for future versions. 

If anyone is interested in moving this further, please write an RFC so that the team can discuss and the community can show interest.

I'm truly sorry this PR has been hanging for so long, and I really appreciate the effort from anyone involved!
